examples/unit/naming/test_naming_advanced.py

```diff
@@ -0,0 +1,123 @@
+import re
+from pathlib import Path
+
+import pytest
+
+from cloud_radar.cf.unit import Stack, Template
+
+
+@pytest.fixture
+def stack():
+    template_path = Path(__file__).parent / "naming_resources.yaml"
+    template = Template.from_yaml(template_path)
+
+    # In these cases I'll usually use a non-existant region to ensure a real region
+    # is not hard coded
+    return template.create_stack(params={"pName": "test"}, region="xx-west-3")
+
+
+def test_naming_conventions(stack: Stack):
+    """
+    This test shows how common naming conventions can be checked across all
+    resources in a stack, including ignoring resource types which do not support
+    names.
+
+    Args:
+            stack (Stack): the stack being used for these test examples
+    """
+
+    # Ideally this dict would be coming from a common library
+    # function that you have shared between all your test cases
+    # (assuming consistency is the goal).
+    #
+    # This defines all the resource types we want to check, the pattern to match,
+    # and either the details of the Tag or Property that the name is held in.
+    type_patterns = {
+        "AWS::EFS::FileSystem": {
+            "Tag": "Name",
+            # This TagProperty is optional. The default is 'Tags',
+            # but some resources use a different property name for
+            # their tags.
+            "TagProperty": "FileSystemTags",
+            "Pattern": r"^[a-z0-9-]*-vol$",
+        },
+        "AWS::S3::Bucket": {
+            "Property": "BucketName",
+            "Pattern": r"^[a-z0-9-]*-xx-west-3[a-z0-9-]*$",
+        },
+        "AWS::S3::BucketPolicy": {
+            # The BucketPolicy type does not support custom names. If we do not
+            # want to set fail_on_missing_type=False when we call the assertion
+            # below then we need to include this type in the dict, and set it
+            # not to be checked.
+            # This approach ensures that types do not slip through unintentionally.
+            "Check": False
+        },
+    }
+
+    stack.assert_resource_type_property_value_conventions(type_patterns)
+
+
+def test_not_matching_pattern(stack: Stack):
+    """
+    This test validates that if a resource is encountered that does not match the
+    defined pattern, the correct assertion error is raised.
+
+    Args:
+            stack (Stack): the stack being used for these test examples
+    """
+
+    type_patterns = {
+        "AWS::EFS::FileSystem": {
+            "Tag": "Name",
+            "TagProperty": "FileSystemTags",
+            "Pattern": r"^[a-z0-9-]*-FAIL$",
+        },
+        "AWS::S3::Bucket": {
+            "Property": "BucketName",
+            "Pattern": r"^[a-z0-9-]*-xx-west-3[a-z0-9-]*$",
+        },
+    }
```

Comment:
This seems like it will not scale very well for a large organization.

Reply:
I think our current configuration (which isn't this exact format) has 50+ entries in it. 

I was trying to avoid making this library need to be aware of where a name for a resource type was held, or if it was having some ability to override it (for where this didn't know where a name was stored). It doesn't help that the AWS documentation for the [name type](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-name.html) doesn't include any resources where the name comes from a Tag. Although, just handling the name property side of the puzzle is better than needing the end user to work this out themselves. Then the configuration could be simplified to:

```python
    type_patterns = {
        "AWS::S3::Bucket": r"^[a-z0-9-]*-xx-west-3[a-z0-9-]*$",
    }
```

Do you have any thoughts on an alternative configuration format (/ sharing approach)?

The approach to sharing this configuration between tests & repositories is the bit I've not really worked out yet - we've got a fair bit of copy/paste between repositories just now but that is not going to be maintainable for long. In a repository that may have a half dozen templates in it we typically have a file `common_test_functions.py` which has a method `get_naming_dict(<some params>)` then each template specific test calls this common method. The initial thinking had been that this test functions file would end up in an internal library that we would share as a dependency between our repositories, but equally could be a configuration file that is shared with submodules or some other approach.